source/plugin/optional/basic.rst

```diff
@@ -0,0 +1,195 @@
+===================
+Optionals Explained
+===================
+
+Much of the Sponge API makes use of Java's `Optional` system on object accessors, but if you've never used ``Optional``
+before this might seem like a bit of a peculiar way of doing things. You might be tempted to ask:
+*"why do I need to perform an extra step when fetching something from an API object?"*
+
+This section gives a brief summary of ``Optional`` and explains how - and perhaps more importantly why -
+it's used throughout the Sponge API.
+
+Let's start with a little history, and look at how accessors - particularly "getters" - typically work when not making
+use of ``Optional``.
+
+1. Implicit Nullable Contracts and Why They Suck
+================================================
+
+Let's say we have a simple API object ``Entity`` with a ``getFoo()`` method which returns the Entity's ``Foo``.
+
+.. image:: /images/optionals1.png
+
+In the olden times of yore, our plugin might fetch and use the ``Foo`` from the entity using the ``getter`` like this:
+
+.. code-block:: java
+
+    public void someEventHandler(Entity someEntity) {
+        Foo entityFoo = someEntity.getFoo();
+        entityFoo.bar();
+    }
+
+The problem arises because - when designing the API - we have to rely an *implicit* contract on the ``getFoo`` method
+with respect to whether the method can (or cannot) return ``null``. This *implicit contract* can be defined in one of
+two ways:
+
+* **In the javadoc** - this is bad because it relies on the plugin author reading the javadoc for the method, and the contract may not be clear to the plugin author
+* **Using nullable annotations** - this is not ideal because in general these annotations require a tool to be of any use, for example relying on the IDE or compiler to handle the annotations.
+
+.. image:: /images/optionals2.png
+
+Let's assume that the ``getFoo()`` method can - as part of its contract - return null. This suddenly means that our
+code above is unsafe as it may result in a ``NullPointerException`` if ``entityFoo`` is null.
+
+.. code-block:: java
+
+    public void someEventHandler(Entity someEntity) {
+        Foo entityFoo = someEntity.getFoo();
+        entityFoo.bar();
+    }
+
+Let's assume our plugin author is savvy to the nullable nature of our ``getFoo`` method and decides to fix the problem
+with null checking. Assuming he has defined a local constant ``Foo``, he resultant code looks like this:
```

Comment:
*the* resultant code (or *his*?). Also, can we use non-gendered pronouns where possible? (ie. he has -> they have).